src/netstandard/WampSharp/WAMP2/V2/Client/Session/WampSessionClient.cs

```diff
@@ -80,93 +95,103 @@ public void Challenge(string authMethod, ChallengeDetails extra)
+                {
+                    mServerProxy.Abort(ex.Details, ex.Reason);
+                    OnConnectionError(ex);
+                }
             }
         }
 
         public void Welcome(long session, WelcomeDetails details)
         {
-            Session = session;
-
-            Interlocked.CompareExchange(ref mIsConnected, 1, 0);
-
-            OnConnectionEstablished(new WampSessionCreatedEventArgs
-                (session, mSentDetails, details, new SessionTerminator(this)));
-            mOpenTask.TrySetResult(true);
+            var (_, action) = ChangeState(StateChangeReason.ReceivedWelcome);
+            switch (action)
+            {
+                case StateChangeAction.RaiseConnectionEstablished:
+                    Session = session;
+
+                    OnConnectionEstablished(new WampSessionCreatedEventArgs
+                        (session, mSentDetails, details, new SessionTerminator(this)));
+                    mOpenTask.TrySetResult(true);
+                    break;
+                case StateChangeAction.SendProtocolViolation:
+                    SendProtocolViolation("Received a WELCOME message from an invalid state");
+                    break;
+            }
         }
 
         public void Abort(AbortDetails details, string reason)
         {
             using (IDisposable proxy = mServerProxy as IDisposable)
             {
-                TrySetCloseEventArgs(SessionCloseType.Abort, details, reason);
+                var (_, action) = ChangeState(StateChangeReason.ReceivedAbort);
+                if (action == StateChangeAction.Close)
+                    TrySetCloseEventArgs(SessionCloseType.Abort, details, reason);
             }
         }
 
         public void Goodbye(GoodbyeDetails details, string reason)
         {
+            var (_, action) = ChangeState(StateChangeReason.ReceivedGoodbye);
             using (IDisposable proxy = mServerProxy as IDisposable)
             {
                 TrySetCloseEventArgs(SessionCloseType.Goodbye, details, reason);
 
-                if (!mGoodbyeSent)
-                {
-                    mGoodbyeSent = true;
-                    mServerProxy.Goodbye(new GoodbyeDetails(), WampErrors.GoodbyeAndOut);
-                }
-                else
+                switch (action)
                 {
-                    mGoodbyeMessage = new GoodbyeMessage(){Details = details, Reason = reason};
+                    case StateChangeAction.SendGoodbye:
+                        mServerProxy.Goodbye(new GoodbyeDetails(), WampErrors.GoodbyeAndOut);
+                        break;
+                    case StateChangeAction.SendProtocolViolation:
+                        SendProtocolViolation("Received GOODBYE message before session was established");
+                        break;
                 }
+                
+                Interlocked.CompareExchange(ref mGoodbyeMessage, new GoodbyeMessage { Details = details, Reason = reason }, null);
```

Comment:
We shouldn't be setting the goodbye message if it's a protocol violation - we want the close task to fail with the abort reason